Source/WebKit/UIProcess/API/APIContentRuleListStore.cpp

```diff
@@ -445,26 +446,38 @@ static WTF::String getContentRuleListSourceFromMappedFile(const MappedData& mapp
     ASSERT(!RunLoop::isMain());
 
     if (mappedData.metaData.version < 9) {
-        // Older versions cannot recover the original JSON source from disk.
+        WTFLogAlways("Content Rule List source recovery failed: Version is too old to recover the original JSON source from disk.");
         return { };
     }
 
-    if (!mappedData.metaData.sourceSize)
+    auto sourceSizeBytes = mappedData.metaData.sourceSize;
+    if (!sourceSizeBytes) {
+        WTFLogAlways("Content Rule List source recovery failed: No source size specified; cannot retrieve content.");
         return { };
+    }
 
+    auto dataSpan = mappedData.data.span();
     auto headerSizeBytes = headerSize(mappedData.metaData.version);
-    bool is8Bit = mappedData.data.span()[headerSizeBytes];
+
+    if (dataSpan.size() < headerSizeBytes + sourceSizeBytes) {
```

Comment:
Should we also fail if it is greater than?

Reply:
Greater than is valid still, since there is more data normally after the original source - if just the version number mismatches.

The other check for total file size catches the other case.